apollo-federation/src/operation/mod.rs

```diff
@@ -104,6 +104,98 @@ fn same_value(left: &executable::Value, right: &executable::Value) -> bool {
+/// duplicates). This is used for hashing input values in a way consistent with [same_value()].
+fn sort_value(value: &executable::Value) -> executable::Value {
+    use apollo_compiler::executable::Value;
+    match value {
+        Value::List(elems) => Value::List(
+            elems
+                .iter()
+                .map(|value| Node::new(sort_value(value)))
+                .collect(),
+        ),
+        Value::Object(pairs) => {
+            let mut pairs = pairs
+                .iter()
+                .map(|(name, value)| (name.clone(), Node::new(sort_value(value))))
+                .collect::<Vec<_>>();
+            pairs.sort_by(|left, right| left.0.cmp(&right.0));
+            Value::Object(pairs)
+        }
+        _ => value.clone(),
+    }
+}
+
+/// Compare sorted input values, which means specifically establishing an order between the variants
+/// of input values, and comparing values for the same variants accordingly. This is used for
+/// hashing directives in a way consistent with [same_directives()].
+///
+/// Note that Floats and Ints are compared textually and not parsed numerically. This is fine for
+/// the purposes of hashing. For object comparison semantics, see [compare_sorted_object_pairs()].
+fn compare_sorted_value(left: &executable::Value, right: &executable::Value) -> std::cmp::Ordering {
+    use apollo_compiler::executable::Value;
+    fn discriminant(value: &Value) -> u8 {
+        match value {
+            Value::Null => 0,
+            Value::Enum(_) => 1,
+            Value::Variable(_) => 2,
+            Value::String(_) => 3,
+            Value::Float(_) => 4,
+            Value::Int(_) => 5,
+            Value::Boolean(_) => 6,
+            Value::List(_) => 7,
+            Value::Object(_) => 8,
+        }
+    }
+    match (left, right) {
+        (Value::Null, Value::Null) => std::cmp::Ordering::Equal,
+        (Value::Enum(left), Value::Enum(right)) => left.cmp(right),
+        (Value::Variable(left), Value::Variable(right)) => left.cmp(right),
+        (Value::String(left), Value::String(right)) => left.cmp(right),
+        (Value::Float(left), Value::Float(right)) => left.as_str().cmp(right.as_str()),
+        (Value::Int(left), Value::Int(right)) => left.as_str().cmp(right.as_str()),
+        (Value::Boolean(left), Value::Boolean(right)) => left.cmp(right),
+        (Value::List(left), Value::List(right)) => left.len().cmp(&right.len()).then_with(|| {
+            left.iter()
+                .zip(right)
+                .map(|(left, right)| compare_sorted_value(left, right))
+                .find(|o| o.is_ne())
+                .unwrap_or(std::cmp::Ordering::Equal)
+        }),
+        (Value::Object(left), Value::Object(right)) => compare_sorted_name_value_pairs(
+            left.iter().map(|pair| (&pair.0, &pair.1)),
+            right.iter().map(|pair| (&pair.0, &pair.1)),
+        ),
+        _ => discriminant(left).cmp(&discriminant(right)),
```

Comment:
@goto-bus-stop 
`left` and `right` can be different `Value` variants, e.g. we're comparing a `String` to `Null`. This can happen due to e.g. needing to sort a directive list representing:
```graphql
@foo(bar: null) @foo(bar: "bar")
```
In particular, if `bar` is a custom scalar, the values compared can be any valid value.

Reply:
Oh, I misunderstood the context this is in, this is to ensure ordering in that case.

Reply:
We might use `std::mem::discriminant` instead of the manual `discriminant()` implementation though: https://doc.rust-lang.org/std/mem/fn.discriminant.html

Reply:
@goto-bus-stop 
It looks like the built-in discriminant doesn't implement `Ord` though, unfortunately.